### PR TITLE
[DNM] hotfix #23768 and hotfix adding index with shard_row_id_bits

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Releated to #23768.

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -1309,6 +1309,17 @@ func (w *worker) sendRangeTaskToWorkers(t table.Table, workers []*addIndexWorker
 		if endKey.Cmp(keyRange.EndKey) < 0 {
 			endIncluded = true
 		}
+		if t.Meta().ShardRowIDBits > 0 {
+			endH, err := getReverseHandle(workers[0].sessCtx.GetStore(), workers[0].priority, t, startHandle, endHandle, endIncluded)
+			if err != nil {
+				logutil.BgLogger().Info("[ddl] send range task to workers, get reverse handle failed", zap.Error(err))
+			} else {
+				logutil.BgLogger().Info("xxx ======================================", zap.Int64("end handle", endHandle), zap.Int64("cur end handle", endH))
```
Comment (crazycs520) [inline ddl/index.go]: duplicate log?

Comment (zimulala) [reply ddl/index.go]: I  removed it in my second commit.

File: ddl/index.go [hunk 2]
```diff
@@ -1613,6 +1624,42 @@ func iterateSnapshotRows(store kv.Storage, priority int, t table.Table, version
...
+	if endIncluded {
+		if endHandle == math.MaxInt64 {
+			return endHandle, nil
+		}
+		endHandle++
+	}
+
+	var handle int64
+	snap, err := store.GetSnapshot(kv.MaxVersion)
+	snap.SetOption(kv.Priority, priority)
```
Comment (crazycs520) [inline ddl/index.go:1637]: move to line1640, should check error first.

File: ddl/index.go [hunk 3]
```diff
@@ -1613,6 +1624,42 @@ func iterateSnapshotRows(store kv.Storage, priority int, t table.Table, version
 	return nil
 }
 
+func getReverseHandle(store kv.Storage, priority int, t table.Table, startHandle, endHandle int64, endIncluded bool) (int64, error) {
+	if endIncluded {
+		if endHandle == math.MaxInt64 {
+			return endHandle, nil
+		}
+		endHandle++
+	}
```
Comment (AilinKid) [inline ddl/index.go:1633]: 如果 endIncluded = true 是不是就是最后那个实体的 key，所有直接 return endHandle, nil 就行了？

File: ddl/index.go [hunk 4]
```diff
@@ -1309,6 +1309,17 @@ func (w *worker) sendRangeTaskToWorkers(t table.Table, workers []*addIndexWorker
 		if endKey.Cmp(keyRange.EndKey) < 0 {
 			endIncluded = true
 		}
+		if t.Meta().ShardRowIDBits > 0 {
+			endH, err := getReverseHandle(workers[0].sessCtx.GetStore(), workers[0].priority, t, startHandle, endHandle, endIncluded)
+			if err != nil {
+				logutil.BgLogger().Info("[ddl] send range task to workers, get reverse handle failed", zap.Error(err))
+			} else {
```
Comment (crazycs520) [inline ddl/index.go:1316]:
```go
			} else if endH < endHandle {
```

File: ddl/index.go [hunk 5]
```diff
@@ -1309,6 +1309,17 @@ func (w *worker) sendRangeTaskToWorkers(t table.Table, workers []*addIndexWorker
 		if endKey.Cmp(keyRange.EndKey) < 0 {
 			endIncluded = true
 		}
+		if t.Meta().ShardRowIDBits > 0 {
```
Comment (AilinKid) [inline ddl/index.go:1312]: 这个限制可以去掉吗，没有 shard bits 下，大量 append 写到最后一个 region 似乎也有同样的问题

File: ddl/index.go (hunk 3 above)
Comment (zimulala) [reply ddl/index.go:1633]: In theory, this is true. In addition, the endHandle and the previous handles may have been deleted. A `get` can also be used

File: ddl/index.go (hunk 5 above)
Comment (crazycs520) [reply ddl/index.go:1312]: 还是先留着这个吧，毕竟是 hotfix 只解决具体问题，别影响原来的其他功能吧。后续在正式版本的修复可以考虑去掉这个限制

File: ddl/index.go (hunk 3 above)
Comment (crazycs520) [reply ddl/index.go:1633]: "如果 endIncluded = true 是不是就是最后那个实体的 key"
-------------------------------------------------------
这个推论不成立吧

File: ddl/index.go (hunk 5 above)
Comment (zimulala) [reply ddl/index.go:1312]: 嗯，我也这么想的

Comment (AilinKid) [reply ddl/index.go:1312]: 好的好的

File: ddl/index.go (hunk 3 above)
Comment (AilinKid) [reply ddl/index.go:1633]: 霜哥怎么理解

Comment (crazycs520) [reply ddl/index.go:1633]: Add following log at the begining of this function:

```go
logutil.BgLogger().Info("getMaxRowID", zap.Int64("end", endHandle), zap.Bool("include", endIncluded))
```

Then execute following SQL:

```sql
create table t (a bigint) shard_row_id_bits=2;
alter table t add index idx1(a);
```

You will found following log:

```log
[2021/06/23 15:08:55.268 +08:00] [INFO] [index.go:1628] [getMaxRowID] [end=-9223372036854775808] [include=true]
```

Here the `endIncluded` is true, but `-9223372036854775808` is not a real key of the table.

**Review by crazycs520 — COMMENTED** [review]
LGTM

**Comment by AilinKid** [comment]
LGTM

File: ddl/index.go (hunk 3 above)
Comment (AilinKid) [reply ddl/index.go:1633]: got it

**Pull request closed without merging**